clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -149,6 +157,74 @@ public Sender(LogContext logContext,
         this.retryBackoffMs = retryBackoffMs;
         this.apiVersions = apiVersions;
         this.transactionManager = transactionManager;
+        this.inFlightBatches = new HashMap<>();
+    }
+
+    public Map<TopicPartition, List<ProducerBatch>> getInflightBatches() {
+        return inFlightBatches;
+    }
+
+    public List<ProducerBatch> inFlightBatches(TopicPartition tp) {
+        return inFlightBatches.containsKey(tp) ? inFlightBatches.get(tp) : new ArrayList<>();
+    }
+
+    public void maybeRemoveFromInflightBatches(ProducerBatch batch) {
+        List<ProducerBatch> batches = inFlightBatches.get(batch.topicPartition);
+        if (batches != null) {
+            batches.remove(batch);
+        }
+    }
+
+    /**
+     *  Get the in-flight batches that has reached delivery timeout.
+     */
+    private List<ProducerBatch> getExpiredInflightBatches(long now) {
+        Set<TopicPartition>  topicPartitions = this.accumulator.getTopicPartitions();
+        List<ProducerBatch> expiredBatches = new ArrayList<>();
+        for (TopicPartition topicPartition : topicPartitions) {
+            List<ProducerBatch> partitionInFlightBatches = inFlightBatches.get(topicPartition);
+            if (partitionInFlightBatches != null) {
+                Iterator<ProducerBatch> iter = partitionInFlightBatches.iterator();
+                while (iter.hasNext()) {
+                    ProducerBatch batch = iter.next();
+                    if (batch.hasReachedDeliveryTimeout(accumulator.getDeliveryTimeoutMs(), now)) {
+                        iter.remove();
+                        // expireBatches is called in Sender.sendProducerData, before client.poll.
+                        // The batch.finalState() == null invariant should always hold. We put an if ... else...
+                        // statement here to log info of the batches that violate the invariant.
```

Comment:
This message needs to be updated.

Reply:
updated the comment: 

                        // expireBatches is called in Sender.sendProducerData, before client.poll.
                        // The batch.finalState() == null invariant should always hold. An IllegalStateException
                        // exception will be thrown if the invariant is violated.